---
title: "GSEA_Weekly_Challenge"
output: html_document
date: "2023-08-07"
editor_options: 
  markdown:
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Hi everyone! We'll be using a subset of the bulk RNA Seq data generated by our lab to do gene set enrichment analysis (GSEA)! This dataset has 4 groups, with 2 samples for each group. There is a pregnant and non - pregnant group, and each group has 2 tissue types, L (lung) and ML (mesenteric lymph node). So, these are the 4 groups we'll be working with: NonPreg_L, NonPreg_ML, Preg_L, and Preg_ML. We have 2 samples for each group, making a total of 8 samples. All of the groups have been infected with Influenza A Virus (IAV). We want to learn more about the gene regulation differences between these groups, and to see how pregnancy affects the immune response to IAV.

Make sure to save the data file (gsea.xlsx) that is needed for this in the same place as this .Rmd file is saved. For example, I could choose to save both this .Rmd file and the data file in my desktop.

Next, we need to make sure we are working out of the correct working directory. This is essentially telling the computer where in the computer system we have saved our data. Go to Session \> Set Working Directory and choose where these files are saved. If I have saved this on my desktop, then I would choose my desktop in 'Set Working Directory'.

A couple things to make this easier, go to Code \> Soft Wrap Long Lines to be able to view the long lines of text. Also, to run the code in the darker gray sections, you can use the little green triangle in the top right corner.

Install the libraries (if you haven't done it in the past) in the console using install.packages() and load in the following data, then take a look at it in the console using View() and head().

```{r load libraries and data}
library(readxl)
library(rjson)
library(tidyr)
library(ggplot2)
gsea_df <- read_xlsx("gsea.xlsx")
gsea_df <- as.data.frame(gsea_df)
rownames(gsea_df) <- gsea_df$gene_name_mmu
```

How many rows and columns does this dataset have? What data do they display? (There is 1 column that has the gene names repeated, ignore that for now) 
Rows: 13368, names of genes 
Columns: 9, number of samples (4 groups with 2 samples for each group)

Next, we need to choose which gene set we are going to do GSEA for and create a heatmap out of. A gene set is a group of genes that are all co-expressed or all members of a biological pathway.

For this exercise, we'll use the 'GOBP_ACTIVATION_OF_INNATE_IMMUNE_RESPONSE' gene set from the Broad Institute. GOBP stands for Gene Ontology: Biological Process.

```{r load gene set}
GOBP <- fromJSON(file = "https://www.gsea-msigdb.org/gsea/msigdb/human/download_geneset.jsp?geneSetName=GOBP_ACTIVATION_OF_INNATE_IMMUNE_RESPONSE&fileType=json")
```

What is the name of the section that has all the gene names/symbols? 
Hint: The answer should follow this format: GOBP$___$___ GOBP$GOBP_ACTIVATION_OF_INNATE_IMMUNE_RESPONSE$geneSymbols

Use the space below to check what type of data it is (numeric, character, logical, integer?) You can use class() to see what type of data it is. If it is not a data frame, change it into a data frame and give it the variable name 'gene_set'.

```{r data frame}
class(GOBP$GOBP_ACTIVATION_OF_INNATE_IMMUNE_RESPONSE$geneSymbols)
gene_set <- as.data.frame(GOBP$GOBP_ACTIVATION_OF_INNATE_IMMUNE_RESPONSE$geneSymbols)
```

Next, we need to check if all the genes that are in the gene set are actually part of our data. One way we can do this is by using the 'merge()' function. This is helpful to merge 2 dataframes by common values. In our case, we will merge our total dataset (gsea_df) with the gene set (gene_set) to see which genes are part of both data sets. 
Hint: This is the syntax for the merge function: merge(df1, df2, by.x = col name that has gene names in df1, by.y = col name that has gene names in df2)

```{r merge df}
common_genes <- merge(gsea_df, gene_set, by.x = "gene_name_mmu", by.y = "GOBP$GOBP_ACTIVATION_OF_INNATE_IMMUNE_RESPONSE$geneSymbols")
```

How many genes are common to both data sets? 190

Now we'll use the column of gene names to make that the row names. Thinking ahead to our heatmap, 190 genes is a lot to look at at once, so let's subset the data so that we're only looking at the first 35 genes (aka first 35 rows) and all the columns. 

```{r row names + subset}
rownames(common_genes) <- common_genes$gene_name_mmu
subset_comgenes <- common_genes[1:35, ]
```

In order to use this data to create a heatmap in ggplot, we need to make it so that the dataframe has condensed columns. Use the gather() function to do this. This function "gathers" all the data and puts it into a condensed number of columns that we specify. Keep the column that has all the gene names, then make 2 new columns that have the sample name and value (number). 
Hint: Make sure to specify which columns from subset_comgenes should be included, as one of the columns is just a list of the gene names
** This part is confusing, let me know if you need some help!

```{r gather}
subset_comgenes <- gather(subset_comgenes, key = "Row", value = "Num", 2:9)
```

We can now use this dataframe (subset_comgenes) to create a heatmap in ggplot to understand more about the gene regulation differences in our data.

Create a ggplot using geom_tile(). Put the samples on the x axis and gene names on the y axis. Set the fill to be equal to the gene expression value. 
Hint: Remember that specifying what we want on the x- and y-axis needs to be specified within aes()

```{r ggplot heatmap}
gsea1 <- ggplot(subset_comgenes, aes(x= Row, y=gene_name_mmu, fill=Num)) +
  geom_tile()
```

Next, let's add onto gsea1 to specify what colors we want to be used for the fill. Typically in GSEA, blue is used for downregulated genes while red is used for upregulated genes. White is used for genes that are not differentially expressed between groups. Use scale_fill_gradient2() to change the coloring.

```{r ggplot heatmap}
gsea2 <- ggplot(subset_comgenes, aes(x= Row, y=gene_name_mmu, fill=Num)) +
  geom_tile() +
  theme_bw() +
  scale_fill_gradient2(
    low = "blue",
    high = "red",
    mid = "white",
    midpoint = 0)
```

Finally, add onto gsea2 by adding a title and x- and y-axis labels. You can label the legend for the color by using fill="l2f change" within labs().

```{r ggplot heatmap}
gsea3 <- ggplot(subset_comgenes, aes(x= Row, y=gene_name_mmu, fill=Num)) +
  geom_tile() +
  theme_bw() +
  scale_fill_gradient2(
    low = "blue",
    high = "red",
    mid = "white",
    midpoint = 0) +
  labs(title = "ACTIVATION_OF_INNATE_IMMUNE_RESPONSE", y = "Genes", x = "Samples", fill="l2f change")
```

Most of the genes look red, meaning they are upregulated. However, one of the genes is downregulated (blue) in most samples, what is the name of this gene? BPIFB1

Look up this gene, if you're correct it should be a gene that plays a part in mucosa of the airways and salivary glands.

YAY you just finished your first GSEA!! :)
